### ISSUE TITLE
URL not found

### ISSUE THREAD

**Issue by 71zenith** [pystardust/ani-cli#560]
**Metadata (please complete the following information)**
Version: ≥ v1.8.1
OS: Arch Linux
Shell: dash
Anime: 86

**Describe the bug**

Prints URL not found.

**Steps To Reproduce**

1. Run `ani-cli -a 10 86`

**Expected behavior**
It should play the video

**Screenshots (if applicable)**
![12:11:03-02-03](https://user-images.githubusercontent.com/92977828/156309552-dd9f822a-887b-4b97-bcaf-34923908f96b.png)

**Additional context**
Problem doesn't persist in 1.8.0. Camrip stuff added in v1.8.1 causes issues. Episodes other than 10 play fine

**Comment by relejek** [comment]
````
Getting data for episode 7 
Downloading episode 7 ...\n 
Exception caught
Exception: [download_helper.cc:451] errorCode=1 Unrecognized URI or unsupported protocol: 

Download failed episode: 007 , please retry or check your internet connection
Currently playing mushoku-tensei-isekai-ittara-honki-dasu-2nd-season-dub episode 7/12
[n] next episode
[p] previous episode
[s] select episode
[r] replay current episode
[a] search for another anime
[h] search history
[q] exit
Enter choice: q
````

yup, not working for me aswell. download of the show dropped halfway, it was fine for first 6eps then this bug appeared at ep 7-12,

**Comment by Derisis13** [comment]
@relejek it appears to me that you're using an old version of ani-cli. Some of your problems might be solved by updating. See how in the readme.

**Comment by relejek** [comment]
Oh, yeah. above error message was before i updated from v1.8.0
now i get same error as @71zenith on v1.8.3

**Comment by asheeeshh** [comment]
same error

**Comment by VG08** [comment]
am also getting same issue

**Comment by parsivan** [comment]
I get the same error with one peace...I forgot what version I was running first but it happened once on ep84 I tried it the episode several times and it eventually loaded but it didn't work for ep85 so I figured I should update but when I patched ani-cli and got to V1.8.3 now none of the episodes of one peace load.
![Screenshot 2022-03-03 6 11 58 AM](https://user-images.githubusercontent.com/21005543/156500619-cf1b0a0d-51fa-45b9-ac7e-5f4adcd4e242.png)

**Comment by parsivan** [comment]
I think this needs immediate attention and priority.

**Comment by noangst** [comment]
Yeah, I'm getting the same Video URL not found error for every video.

**Comment by Hallos74** [comment]
im gettin the same problem.
it was working fine yesterday tho

**Comment by 71zenith** [comment]
The error y'll are getting is due to a different reason

**Comment by alpha-hexor** [comment]
site has changed
just a normal curl request to the embade_url giving this link [death note as example]
```bash
sources:[{file: 'https://hlsx02.asiancdn.net/astream/92bd0b67e03416084ffd8bb811b736ff/death-note-episode-1.1633168393.m3u8',label: 'hls P','type' : 'hls'},]
```
we have to use this link i guess

**Comment by VG08** [comment]
Which is?

**Comment by Mstkj** [comment]
Same issue for me. Running v1.8.3 on Arch with zsh.

**Comment by matthewmiaow** [comment]
Same for me -- the error appears on all episodes.

**Comment by ura43** [comment]
Just for the record, this was a small parsing error yesterday.
The script is now broken due to increased gogoanime security.
Since you have all made your way here, we're repurposing the issue for that

**Comment by kalix123** [comment]
![03-03-22-02_43_19](https://user-images.githubusercontent.com/8263176/156519464-d7b30f1b-50c9-477a-9f79-e5e1d1c29add.png)
Inspecting the message you can still get a link to the stream (or the embed that is). For this example, the link would be,
[](https://gogoplay4.com/streaming.php?id=MTA2MjQ=&title=Death+Note+Episode+1)

**Comment by ura43** [comment]
Running `bash -x ani-cli flcl` and selecting `flcl-dub`, I get the following relevant debug info:
```
+ dpage_link='https://gogoplay4.com/embedplus?id=NzM1MTM=&token=xJtjz5PCAAuREq3PGViG9w&expires=1646295421'
++ get_video_quality 'https://gogoplay4.com/embedplus?id=NzM1MTM=&token=xJtjz5PCAAuREq3PGViG9w&expires=1646295421'
++ dpage_url='https://gogoplay4.com/embedplus?id=NzM1MTM=&token=xJtjz5PCAAuREq3PGViG9w&expires=1646295421'
+++ decrypt_link 'https://gogoplay4.com/embedplus?id=NzM1MTM=&token=xJtjz5PCAAuREq3PGViG9w&expires=1646295421'
+++ ajax_url=https://gogoplay.io/encrypt-ajax.php
+++ secret_key=3235373436353338353932393338333936373634363632383739383333323838
+++ iv=34323036393133333738303038313335
++++ printf %s 'https://gogoplay4.com/embedplus?id=NzM1MTM=&token=xJtjz5PCAAuREq3PGViG9w&expires=1646295421'
++++ cut '-d?' -f2
++++ cut '-d&' -f1
++++ sed s/id=//g
+++ video_id=NzM1MTM=
++++ printf %s NzM1MTM=
++++ openssl enc -aes256 -K 3235373436353338353932393338333936373634363632383739383333323838 -iv 34323036393133333738303038313335 -a
+++ ajax=bYiO5MUWIjqSPbNcQO/aVQ==
+++ curl -s -H x-requested-with:XMLHttpRequest https://gogoplay.io/encrypt-ajax.php -d id=bYiO5MUWIjqSPbNcQO/aVQ== -d time=69420691337800813569
+++ sed -e 's/\].*/\]/' -e 's/\\//g'
+++ grep -Eo 'https:\/\/[-a-zA-Z0-9@:%._\+~#=][a-zA-Z0-9][-a-zA-Z0-9@:%_\+.~#?&\/\/=]*'
++ video_links=
++ case $quality in
+++ printf %s ''
+++ head -n 4
+++ tail -n 1
++ video_link=
++ printf %s ''
+ video_url=
+ '[' 0 -eq 1 ']'
+ '[' 0 -eq 0 ']'
+ sed -E '
			s/^flcl-dub\t[0-9]+/flcl-dub\t2/
		' /home/ura/.cache/ani-hsts
+ '[' 0 -ne 0 ']'
+ '[' -z '' ']'
+ die 'Video URL not found'
+ err 'Video URL not found'
+ printf '\033[1;31m%s\033[0m\n' 'Video URL not found'
Video URL not found
+ exit 1
```

**Comment by ura43** [comment]
> site has changed just a normal curl request to the embade_url giving this link [death note as example]
> 
> ```shell
> sources:[{file: 'https://hlsx02.asiancdn.net/astream/92bd0b67e03416084ffd8bb811b736ff/death-note-episode-1.1633168393.m3u8',label: 'hls P','type' : 'hls'},]
> ```
> 
> we have to use this link i guess

That link doesn't work for me, at least not ootb. Could you provide further context?

**Comment by alpha-hexor** [comment]
Yea the these links aren't working

**Comment by alpha-hexor** [comment]
0

> ![03-03-22-02_43_19](https://user-images.githubusercontent.com/8263176/156519464-d7b30f1b-50c9-477a-9f79-e5e1d1c29add.png) Inspecting the message you can still get a link to the stream (or the embed that is). For this example, the link would be, https://gogoplay4.com/streaming.php?id=MTA2MjQ=&title=Death+Note+Episode+1

Can we use burp suite for that??

**Comment by kalix123** [comment]
From my little bit of testing it seems that the major problem now is just finding that vidstreamingcdn link.
I was able to go to gogoanime inspect the player and get a link https://gogoplay4.com/streaming.php?id=MTA2MjQ=&title=Death+Note+Episode+1(this one doesn't work anymore since the id keeps changing). The from there I inspected the jw player and searched the html for "mp4" and found this link ""https://vidstreamingcdn.com/cdn5/92bd0b67e03416084ffd8bb811b736ff/EP.1.v0.1639176238.720p.mp4?mac=NoccSKJIJs8pLnDW5Riau4HBdJ8UpKr8mtUZEf96i6k%3D&vip=2601:548:c101:7090::157d&expiry=1646301028974"
Which worked when using with mpv and the referrer as one normally would.

**Comment by kalix123** [comment]
Steps that I took to get the vidstreaming link
step 1.
![03-03-22-03_05_31](https://user-images.githubusercontent.com/8263176/156523445-119c333b-1aaa-45eb-9f9c-f5de19e474f9.png)
Go to gogoanime/the video and find the streaming.php link (as pictured)

step 2.
open the link in a browser other then chrome (for me if I do it in chrome, opening dev tools will redirect the page somewhere else).

step 3
open dev tools by pressing f12 at least that's what it is in firefox

step 4
start playing the video then search the html for "mp4"

step 5
it looks like this:
![03-03-22-03_08_49](https://user-images.githubusercontent.com/8263176/156523416-daf72199-1c4c-4137-8521-df077aae52ab.png)

**Comment by ura43** [comment]
@alpha-hexor we're currently burpsuiting the thing. Feel free to join us if you have discord

**Comment by ura43** [comment]
https://developers.google.com/recaptcha/docs/v3

**Comment by rbjnazz** [comment]
``` bash
❯ ani-cli -V
Version: 1.8.0 
❯ ani-cli -q 480 -d "one piece"
[1] one-piece
[2] toriko-dub
[3] one-piece-3d2y
[4] one-piece-dub
[5] one-piece-film-z
[6] one-piece-special
[7] one-piece-nami-ova
[8] one-piece-3d2y-dub
[9] one-piece-film-gold
[10] one-piece-film-z-dub
[11] one-piece-heart-of-gold
[12] hungry-days-aoharu-ka-yo
[13] one-piece-film-gold-dub
[14] one-piece-movie-14-stampede
[15] one-piece-straw-hat-theater
[16] one-piece-3d-mugiwara-chase
[17] one-piece-episode-of-sorajima
[18] one-piece-episode-of-east-blue
[19] one-piece-movie-10-strong-world
[20] one-piece-adventure-of-nebulandia
[q] exit
Enter choice: 1
Range of episodes can be specified: start_number end_number 
Choose episode[1-1012]: 851

Getting data for episode 851 
Current video quality is not available (defaulting to best quality)
Downloading episode 851 ... 
Exception caught
Exception: [download_helper.cc:451] errorCode=1 Unrecognized URI or unsupported protocol: 

Download failed episode: 851 , please retry or check your internet connection

```

**Comment by alpha-hexor** [comment]
New ajax parameters found
![burp](https://user-images.githubusercontent.com/83006189/156544067-dbb09198-7209-485a-a539-bc7be89d3405.png)

**Comment by ura43** [comment]
> New ajax parameters found ![burp](https://user-images.githubusercontent.com/83006189/156544067-dbb09198-7209-485a-a539-bc7be89d3405.png)

Wanna add me on tryhackme?

**Comment by alpha-hexor** [comment]
Yea but kinda inactive

**Comment by ura43** [comment]
https://github.com/justfoolingaround/animdl/commits/master

animdl beat us to it

**Comment by CoolnsX** [comment]
working on it

**Comment by CoolnsX** [comment]
@alpha-hexor can you help me how to convert those gogoanime keys and ivs.. like u did last time.. I am not so gud in python..

**Comment by alpha-hexor** [comment]
yea sure working on it

**Comment by alpha-hexor** [comment]
the decrypt function will be like this
```bash
echo $data | base64 -d | openssl enc -d -aes256  -K $secret -iv $iv
```
then return the output

**Comment by alpha-hexor** [comment]
did you created the encrypt function??

**Comment by CoolnsX** [comment]
> did you created the encrypt function??

encrypt fn??

**Comment by ura43** [comment]
Yes, you now have to encrypt at least on parameter for later use

**Comment by ura43** [comment]
@alpha-hexor can you make a proof of concept for the encrypt function?

**Comment by alpha-hexor** [comment]
we have to create the encrypt function too
the rough algo is like this
```bash
#encrypt_fn(some data)
#decrypt_fn(some data)

#get the crypto value from this  data-value="bSlDTEA48388xUL8eGboKkzwqNnLdSeRzd3hYp/lax1pxCfcTgcmKZEM0bbDF3eTrjqXWUyid3jTFHje356qJKgNye9fne/K7QGYqyWOvzO5Jem8OQItdesl4GbQ6vSk"

#assume 
#x=cypto_value
#then steps will be like this
content_info = decrypt_fn(x)
new_content_info = echo $content_info | cut -d '&' -f1
id = encrypt_fn(new_content_info)

response = curl 'the headers' $ajax_url -d id=id

final_response = decrypt_fn(response)

```

**Comment by alpha-hexor** [comment]
As the encrytp function only used once so no need to make a function of it
```bash
#taking reference from above algo
pad_data="\x08\x0e\x03\x08\t\x03\x04\t"
new_payload=${new_content_info}${pad_data}
id=$(echo -n -e $new_payload | openssl enc -aes256  -K $secret -iv $iv -a )
response = curl 'the headers' $ajax_url -d id=id
```

**Comment by CoolnsX** [comment]
> we have to create the encrypt function too the rough algo is like this
> 
> ```shell
> #encrypt_fn(some data)
> #decrypt_fn(some data)
> 
> #get the crypto value from this  data-value="bSlDTEA48388xUL8eGboKkzwqNnLdSeRzd3hYp/lax1pxCfcTgcmKZEM0bbDF3eTrjqXWUyid3jTFHje356qJKgNye9fne/K7QGYqyWOvzO5Jem8OQItdesl4GbQ6vSk"
> 
> #assume 
> #x=cypto_value
> #then steps will be like this
> content_info = decrypt_fn(x)
> new_content_info = echo $content_info | cut -d '&' -f1
> id = encrypt_fn(new_content_info)
> 
> response = curl 'the headers' $ajax_url -d id=id
> 
> final_response = decrypt_fn(response)
> ```

from where we get that long data-value??
also i tried to decrypt the above data value and checked it against the original value.. they doesn't have that much difference.. only character.. in my case

id from site=MTA2MjQ
decrypting above data-value=MTA2MjQ7

**Comment by alpha-hexor** [comment]
the data value can be found like this

first

![image](https://user-images.githubusercontent.com/83006189/156606206-a79d025b-58ff-487b-b55a-d6cbc0a5c3b4.png)

upon visiting the selected website u can find the data value from this line

![image](https://user-images.githubusercontent.com/83006189/156606517-0f482008-24c6-4953-b526-83daa69ff80c.png)

**Comment by alpha-hexor** [comment]
> > we have to create the encrypt function too the rough algo is like this
> > ```shell
> > #encrypt_fn(some data)
> > #decrypt_fn(some data)
> > 
> > #get the crypto value from this  data-value="bSlDTEA48388xUL8eGboKkzwqNnLdSeRzd3hYp/lax1pxCfcTgcmKZEM0bbDF3eTrjqXWUyid3jTFHje356qJKgNye9fne/K7QGYqyWOvzO5Jem8OQItdesl4GbQ6vSk"
> > 
> > #assume 
> > #x=cypto_value
> > #then steps will be like this
> > content_info = decrypt_fn(x)
> > new_content_info = echo $content_info | cut -d '&' -f1
> > id = encrypt_fn(new_content_info)
> > 
> > response = curl 'the headers' $ajax_url -d id=id
> > 
> > final_response = decrypt_fn(response)
> > ```
> 
> from where we get that long data-value?? also i tried to decrypt the above data value and checked it against the original value.. they doesn't have that much difference.. only character.. in my case
> 
> id from site=MTA2MjQ decrypting above data-value=MTA2MjQ7

the final encrypted data will differ as we have to use some data padding

**Comment by alpha-hexor** [comment]
finally u should get the result like this

![image](https://user-images.githubusercontent.com/83006189/156608895-3baa694d-1477-4c1c-a93d-6d466e41f8fe.png)

then decrypting the data part will give us the final links

![image](https://user-images.githubusercontent.com/83006189/156609176-2a56f973-181b-44ce-a48b-64e4b65ebf3c.png)

**Comment by CoolnsX** [comment]
seems to be having problem in padding link.. rest before is working properly..

**Comment by alpha-hexor** [comment]
try this one

https://gist.github.com/alpha-hexor/1050c464e720b7b91e86e1b97c6b684a

**Comment by alpha-hexor** [comment]
> seems to be having problem in content info with padding.. rest before is working properly..
> 
> ```
> decrypt_link() {
>     pad_data="\x08\x0e\x03\x08\t\x03\x04\t"
>     ajax_url="https://gogoplay4.com/encrypt-ajax.php"
>     secret_key='3235373136353338353232393338333936313634363632323738383333323838'
>     iv='31323835363732393835323338333933'
>     content_info=$(printf "$1" | base64 -d | openssl enc -d -aes256 -K "$secret_key" -iv "$iv" | cut -d\& -f1)
>     new_payload=${content_info}${pad_data}
>     id=$(echo -n -e "$new_payload" | openssl enc -aes256  -K $secret -iv $iv -a )
>     result=$(curl -s -H "X-Requested-With:XMLHttpRequest" -d "id=$id" "https://gogoplay4.com/encrypt-ajax.php")
>     printf "$result" | base64 -d | openssl enc -d -aes256 -K "$secret" -iv "$iv"
> }
> ```
> 
> the id after encoding are like this `���˦0����c)1�b�T"gF����_8�`
> 
> problem is I am unable to get the data after getting this id

think result will be something like this ["data":"....."] we only have to decrypt the "..." part i think that's why

**Comment by CoolnsX** [comment]
yup i found the problem.. and trying to replace python with something easy..

**Comment by FireTonnor** [comment]
How can i fix it can someone please help me

**Comment by ZastianPretorius** [comment]
> How can i fix it can someone please help me

They are working on it. Should be working soon.

**Comment by FireTonnor** [comment]
Oh ok

**Comment by parsivan** [comment]
Can someone explain what went wrong? I want to learn from this. :) If it's no too much to ask!
Trying to learn Linux bash scripting and terminal! Basically a noob :/

**Comment by ura43** [comment]
> Can someone explain what went wrong? I want to learn from this. :) If it's not too much to ask! Trying to learn Linux bash scripting and terminal! Basically a noob :/ I heard someone say it's because the website upped their security game XD.

tldr: if you want to learn shell scripting, you don't wanna know what happened here.
Now if you want to learn web hacking: gogoanime changed keys and heavily obfuscated their player and their site in general

**Comment by asheeeshh** [comment]
Would using some other site won't fix it, though I doubt it would lead to changing the whole script maybe, a lot of unnecessary work ig.

**Comment by parsivan** [comment]
> > Can someone explain what went wrong? I want to learn from this. :) If it's not too much to ask! Trying to learn Linux bash scripting and terminal! Basically a noob :/ I heard someone say it's because the website upped their security game XD.
> 
> tldr: if you want to learn shell scripting, you don't wanna know what happened here.
> Now if you want to learn web hacking: gogoanime changed keys and heavily obfuscated their player and their site in general

yeah i wanted to know about security aspects too. thanks 👍
